target/linux/realtek/files-6.12/drivers/net/dsa/rtl83xx/rtl931x.c

```diff
@@ -791,11 +795,30 @@ static u64 rtl931x_read_l2_entry_using_hash(u32 hash, u32 pos, struct rtl838x_l2
 
 static u64 rtl931x_read_cam(int idx, struct rtl838x_l2_entry *e)
 {
-	return 0;
+	u32 r[4];
+	struct table_reg *q = rtl_table_get(RTL9310_TBL_0, 1);
+	rtl_table_read(q, idx);
+	for ( int i = 0; i < 4; i++)
+		r[i] = sw_r32(rtl_table_data(q, i));
+
+	rtl_table_release(q);
+	rtl931x_fill_l2_entry(r, e);
+	if (!e->valid)
+		return 0;
+
+	return ((((u64)(r[0] & 0xffff) << 32) | (u64)r[1]) << 12) | e->vid;
 }
 
 static void rtl931x_write_cam(int idx, struct rtl838x_l2_entry *e)
 {
+	u32 r[4];
```

Comment:
@0xharshal here you create (correctly) a 4x 32 bit array. But `rtl931x_fill_l2_row` only initializes 3x 32 bit (for valid and != valid). As result, the `r[3]` is never initialized. It will then simply write write whatever is currently on the stack in this place in the  loop:

```
for ( int i = 0; i < 4; i++)
		r[i] = sw_r32(rtl_table_data(q, i));
```

We must assume that the data on the stack is basically random. This affects:

*`rtl931x_write_cam` (`RTL9310_TBL_0, 1`)
* `rtl931x_write_l2_entry_using_hash` (`TL9310_TBL_0, 0`)


I tried to have a quick look at https://svanheule.net/realtek/mango/register/tbl_access_ctrl_0 but it seems like the `r[3]` (or bit `31:0`) has various purposes and I can't say for sure what will happen if random data is written to it